packages/flutter/lib/src/foundation/logging.dart

```diff
@@ -0,0 +1,103 @@
+/// as JSON strings using `json.encode()`. In the event that logging is not
+/// enabled for the given [channel], [messageCallback] will not be evaluated.
+/// The cost of logging calls can be further mitigated at call sites by invoking
+/// them in a function that is only evaluated in profile or debug modes. For
+/// example,
+///
+/// ```dart
+/// profile(() {
+///   debugLogEvent(logGestures, () => <String, int> {
+///    'x' : x,
+///    'y' : y,
+///    'z' : z,
+///   });
+/// });
+///```
+///
+/// ignores logging entirely in release mode and no performance penalty is paid.
+///
+/// Logging for a given event channel can be enabled programmatically via
+/// [debugEnableLogging] or using a VM service call.
+///
+void debugLogEvent(
+    LoggingChannel channel, DebugLogMessageCallback messageCallback) {
+  assert(channel != null);
+  if (!debugShouldLogEvent(channel)) {
+    return;
+  }
+
+  assert(messageCallback != null);
+  final Object message = messageCallback();
+  assert(message != null);
+
+  developer.log(json.encode(message), name: channel.name);
+}
+
+final Set<LoggingChannel> _enabledEventChannels = Set<LoggingChannel>();
+
+/// All logging event channels.
+List<LoggingChannel> get debugLogEventChannels =>
+    LoggingChannel._channels.values.toList(growable: false);
+
+/// The set of all enabled logging event channels.
+Set<LoggingChannel> get enabledDebugLogEventChannels => _enabledEventChannels;
+
+/// Identifies a logging channel.
+class LoggingChannel {
+  static final Map<String, LoggingChannel> _channels =
+      <String, LoggingChannel>{};
+
+  /// A uniquely identifying stream name.
+  final String name;
+
+  /// An optional description, suitable for presentation by tools.
+  final String description;
+
+  /// Singleton constructor. Calling `LoggingChannel(name)` returns the same
+  /// actual instance whenever it is called with the same string name.
+  factory LoggingChannel(String name, {String description}) =>
+      _getOrRegisterChannel(name, description ?? '');
+
+  const LoggingChannel._(this.name, this.description);
+
+  static LoggingChannel _getOrRegisterChannel(String name, String description) {
+    assert(name != null);
+    return _channels.putIfAbsent(
+        name, () => LoggingChannel._(name, description));
+  }
+}
+
+/// Get the logging channel registered to this name, or null if none exists.
+LoggingChannel getRegisteredChannel(String name) =>
+    LoggingChannel._channels[name];
+
+/// Enable (or disable) logging for all events on the given [channel].
+void debugEnableLogging(LoggingChannel channel, [bool enable = true]) {
+  assert(channel != null);
+  if (enable) {
+    _enabledEventChannels.add(channel);
+  } else {
+    _enabledEventChannels.remove(channel);
+  }
+}
+
+/// Returns true if events on the given event [channel] should be logged.
+bool debugShouldLogEvent(LoggingChannel channel) {
+  assert(channel != null);
+  return _enabledEventChannels.contains(channel);
+}
```

Comment:
There's a lot of machinery here to support the case of asking for all the registered channels. However, I don't see anything in the use case doc that requires the ability to ask for all registered channels. The mechanisms discussed all seem to be specific to one channel, which it can then assume exists (since there's no difference between it not existing and it existing but sending no data). It's not clear to me what one would do with the list.

Having these mechanisms has a number of problems:

* Dart has no initialization mechanism, so there's no way to know when to query the list.
* The list could be changed at any time, since there's no mechanism to refuse new channels. That means that the service protocol needs a mechanism to announce when the list changes.
* There's no mechanism to remove a channel once it's no longer relevant (e.g. after a hot reload that removes a dependency on a package or plugin that introduces a logging channel).

Reply:
Sorry, I added a use case in the PR description.  The motivation for providing a list of all channels is so that tools can present that list to developers who can pick what to enable.  A description of the channel could further help them decide.

Thanks for identifying some issues.  A few thoughts:

> Dart has no initialization mechanism, so there's no way to know when to query the list.

Right.  This sucks and a real potential headache.

> The list could be changed at any time, since there's no mechanism to refuse new channels. That means that the service protocol needs a mechanism to announce when the list changes.

Agreed.  Adding a notification mechanism is something @devoncarew suggested too.

> There's no mechanism to remove a channel once it's no longer relevant (e.g. after a hot reload that removes a dependency on a package or plugin that introduces a logging channel).

Good point.  I guess I'm not imaging so many channels that this would be a huge problem in practice but you're right there's something unhygienic about no cleanup.

Reply:
Providing a list of all channels is so that tools can present that list to developers who can pick what to enable doesn't make any sense to me. We have a very specific set of logging channels in mind with very specific visualisations. The data isn't human-readable in the general case. I don't really understand the use case for the user looking at a list and choosing from it.

Reply:
A few thoughts.

> We have a very specific set of logging channels in mind with very specific visualisations.

Hmmmm.  I'm skeptical that this won't evolve over time and I can tell you from painful experience managing code on the client side that bakes in logic is conditional on `flutter` version is a drag.

And besides, this misses the opportunity to provide a good experience for package-contributed channels.  Suppose I use the `redux` package, for example.  How would I know what logging channels it supports?

As for specific visualizations, the goal is to support structure in log messages on the client side in a fairly general way.  Certain channels (or conventions) may earn special treatment but the general case should present uniformly well.

> I don't really understand the use case for the user looking at a list and choosing from it.

Turning this around, how are you envisioning a tool would know what channels are available for a given app if it can't ask it?  And how would users know what kind of logging data is available to them without a list to choose from?

Reply:
If you use the redux package and use the redux extension that knows how to show the redux package's logging then presumably they would agree on what channel to listen to without user involvement.

> how are you envisioning a tool would know what channels are available for a given app if it can't ask it?

I wasn't envisioning that at all. The use case document lists five use cases, and to the extend that the use case doc mentions UX, it's that there's a dedicated toggle for each one. We don't need any discovery mechanism at all for what that document describes.

Reply:
> If you use the redux package and use the redux extension

I'm not sure you mean by the redux extension.  Is this a plugin to the IDE or something else?  Where would it live?  I think basic support for structured messages might make any special awareness of redux on the IDE side unnecessary.

> If you use the redux package and use the redux extension that knows how to show the redux package's logging

As I mentioned above I think this supposes a tighter coupling between IDE and framework than is necessary.  Moreover, it's inviting maintenance headaches if we ever want to add or modify logging channels since we'd have to introduce a bunch of conditional logic on the IDE side.  A self-describing framework seems like an easy way around this.

> I wasn't envisioning that at all. The use case document lists five use cases, and to the extent that the use case doc mentions UX, it's that there's a dedicated toggle for each one. We don't need any discovery mechanism at all for what that document describes.

Right.  Sorry.   I didn't think the conversation was limited to that doc.  Since it is not generally readable (e.g., by folks possibly following along on this PR) I didn't think to back-port ideas from here assuming that this conversation was the right place to evolve motivations.  I'll go ahead and migrate the use cases from this PR's description back to it to keep things consistent.